cilium-cli/connectivity/check/scenario.go

```diff
@@ -30,7 +33,32 @@ type ConditionalScenario interface {
 }
 
 type ScenarioBase struct {
+	filepath string
 }
 
 func NewScenarioBase() ScenarioBase {
+	return ScenarioBase{
+		filepath: getSourceFile(),
+	}
+}
+
+func (s ScenarioBase) FilePath() string {
+	return s.filepath
+}
+
+func getSourceFile() string {
+	// 2 steps up go to NewScenarioBase() => actual scenario constructor.
+	_, path, _, ok := runtime.Caller(2)
+	if ok {
+		// 'path' is an absolute path on disk. Trim back to a relative
+		// path from the root directory of the repository, calculated
+		// using this filepath's relationship with the root directory.
+		// If you move this logic, ensure that this calculation directs
+		// back up to the root of the tree where CODEOWNERS exists!
+		_, thisPath, _, _ := runtime.Caller(0)
+		repoDir, _ := filepath.Abs(filepath.Join(thisPath, "..", "..", "..", ".."))
+		return strings.TrimPrefix(path, repoDir+string(filepath.Separator))
+	}
+	// Fall back to the general owner of connectivity infrastructure.
+	return "cilium-cli/connectivity/"
```

Comment:
What is this doing? Is it getting the source file where the test lives?

Reply:
Yeah, it walks up the callstack two levels (to find the function corresponding to the test declaration) and then pulls the path of the file at that callstack level. This is then stored with the test scenario for later lookup of the codeowners structure.